### PR TITLE
Implement predictive test selection model training, cutoffs calibration, and PR tests prediction

### LINKED ISSUE #10506: Training Classifier
After we obtained the preprocessed data from #10505, we need to train a classifier that, given a PR and a test it would output the probability of this test failing on the respective PR.

1. Decide how much data we will gather (i.e., how many PRs we will extract data from) (This will need experimenting with different amounts of data and monitoring the performance of the classifier, it will be a trade-off between model performance and time for data extraction and training).
2. Split the gathered data into training and testing sets. (Check work done in #10505)
3. Train different classifiers on the training data (e.g., Logistic Regression, Decision Tree, Random Forest, XGBoost) and evaluate their performance on the testing data using accuracy, precision, recall, and F1-score, and choose the best classifier for the task.
4. Experiment with feature selection and preprocessing techniques to improve model performance.
5. Train the best classifier on the whole data. (Not sure if we should do this since we won't be able to measure the performance of the resulting classifier, need to read if this is correct)
6. Export the trained classifier to be used in the test failure probability prediction program.

### LINKED ISSUE #10517: Test Failure Prediction Program
This is the program that will be used in production when a new PR is created, and we want to predict the secondary tests that have the highest probability of failing on this PR. The input given to the program will be the PR number (e.g., 10490)

1. We will take the output obtained from the second mode in #10510
2. Preprocess it using the script developed in #10505
3. Input the preprocessed data to the extracted trained classifier from #10506 to predict the probability of failure of each test on the given PR.

   Which will output a list similar to this:
   ```
   Test Name | Predicted Failure Probability
    ----------------------------------------
    test_a   | 92%
    test_b   | 87%
    test_c   | 79%
    test_d   | 4%
    ```

4. We will need to find a way to efficiently parameter tune `ScoreCutOff` and `CountCutOff` based on the desired values for `ChangeRecall`, `TestRecall`, and `SelectionRate`.

> [!WARNING]
> This part is a bit tricky, since we want to experiment with different values for the `CutOff` parameters and need a testing set where the full test suite has run on each PR, not just a subset of selected tests. That way, we can properly calculate quality metrics and choose the best `CutOff` values.
We'll have such a testing set the first time the model is deployed, since up to that point, all tests in the suite have been running on PRs. However, if we go with the idea of retraining the model periodically, like the approach discussed in [Facebook’s paper](https://arxiv.org/pdf/1810.05286) (sections IV-C and IV-D), then future retraining rounds will only have PRs that ran a subset of tests selected by the model. In that case, we won’t be able to calculate quality metrics accurately. So we **should first decide whether we want to adopt periodic retraining**, and if we do, we’ll also need to design a proper workflow for it.

> [!NOTE]
> There might be some way to use the Jenkins tests that run periodically on the whole codebase to help with performance monitoring and the above retraining workflow.

5. After successfully obtaining the suitable `CutOff` parameters, we can apply these thresholds on the test failure probability list to obtain the subset of tests that we will run on the given PR and output them as a CSV file containing the test names:

   e.g., `test_a, test_b, test_c`

### PR DESCRIPTION
**Pull request by akayiz**
## What does this PR change?

Closes #10506
Closes #10517

## Codespace

Check if you already have a running container clicking on [Running CodeSpace](https://github.com/codespaces)

[Create CodeSpace](https://codespaces.new/uyuni-project/uyuni)  [About billing for Github Codespaces](https://docs.github.com/en/billing/managing-billing-for-github-codespaces/about-billing-for-github-codespaces) [CodeSpace Billing Summary](https://github.com/settings/billing/summary) [CodeSpace Limit](https://github.com/settings/billing/spending_limit)

## GUI diff

No difference.

- [x] **DONE**

## Documentation
- In README.md

- [x] **DONE**

## Test coverage
ℹ️ If a major new functionality is added, it is **strongly recommended** that tests for the new functionality are added to the Cucumber test suite
- No tests

- [x] **DONE**

## Links

Issue(s): #10506, #10517

- [x] **DONE**

## Changelogs

Make sure the changelogs entries you are adding are compliant with https://github.com/uyuni-project/uyuni/wiki/Contributing#changelogs and https://github.com/uyuni-project/uyuni/wiki/Contributing#uyuni-projectuyuni-repository

If you don't need a changelog check, please mark this checkbox:

- [x] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

# Before you merge

Check [How to branch and merge properly](https://github.com/uyuni-project/uyuni/wiki/How-to-branch-and-merge-properly)!

### CODE REVIEW

File: testsuite/ext-tools/predictive-test-selection/README.md [hunk 1]
```diff
@@ -360,4 +372,170 @@ Additionally, failed features/tests are underrepresented in our data. To try to
...
+
+`change_recall_heatmap.png`
+
+![](images/change_recall_heatmap.png)
+
+The calibration process generates several visualization files in a `visualizations/` folder to help understand the performance trade-offs. These visualizations help you understand how different cutoff settings affect performance and make informed decisions about target metrics.
+
+#### Prediction Mode Output
+
+Prediction Results JSON: Contains selected tests with failure probabilities
```
Comment (srbarrios) [inline testsuite/ext-tools/predictive-test-selection/README.md]: It will be good if you can add here, in addition to the output, what was the command and arguments executed and the input provided.

Comment (akayiz) [reply testsuite/ext-tools/predictive-test-selection/README.md]: Done. I really appreciate you pointing it out.
